# Exploratory Data Analysis - Bengali AI Dataset
# Grapheme Combinations

This dataset contains images of individual hand-written [Bengali characters](https://en.wikipedia.org/wiki/Bengali_alphabet). 
Bengali characters (graphemes) are written by combining three components: a grapheme_root
, vowel_diacritic, and consonant_diacritic. Your challenge is to classify the components of the grapheme in each
image. There are roughly 10,000 possible graphemes, of which roughly 1,000 are represented in the training set. The
test set includes some graphemes that do not exist in train but has no new grapheme components. It takes a lot of
volunteers filling out [sheets like this](https://github.com/BengaliAI/graphemePrepare/blob/master/collection/A4/form_1.jpg)
to generate a useful amount of real data; focusing the problem on the grapheme components rather than on recognizing
whole graphemes should make it possible to assemble a Bengali OCR system without handwriting samples for all 10,000
graphemes.

In [1]:
import os
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.metrics import confusion_matrix
from IPython.display import Markdown, HTML
# from ..code.jupyter import grid_df_display, combination_matrix

pd.set_option('display.max_columns',   500)
pd.set_option('display.max_colwidth',   -1 if os.environ.get('KAGGLE_KERNEL_RUN_TYPE','') else None)

%load_ext autoreload
%autoreload 2

## Custom Library Functions

In [2]:
# Source: https://stackoverflow.com/questions/38783027/jupyter-notebook-display-two-pandas-tables-side-by-side/50899244#50899244
import pandas as pd
from IPython.display import display,HTML

def grid_df_display(list_dfs, rows = 2, cols=3, fill = 'cols'):
    if fill not in ['rows', 'cols']: print("grid_df_display() - fill must be one of: 'rows', 'cols'")

    html_table = "<table style='width:100%; border:0px'>{content}</table>"
    html_row   = "<tr style='border:0px'>{content}</tr>"
    html_cell  = "<td style='width:{width}%;vertical-align:top;border:0px'>{{content}}</td>"
    html_cell  = html_cell.format(width=100/cols)

    cells = [ html_cell.format(content=df.to_html()) for df in list_dfs[:rows*cols] ]
    cells += cols * [html_cell.format(content="")] # pad

    if fill == 'rows':   # fill in rows first (first row: 0,1,2,... col-1)
        grid = [ html_row.format(content="".join(cells[i:i+cols])) for i in range(0,rows*cols,cols)]
    elif fill == 'cols': # fill columns first (first column: 0,1,2,..., rows-1)
        grid = [ html_row.format(content="".join(cells[i:rows*cols:rows])) for i in range(0,rows)]
    else:
        grid = []

    # noinspection PyTypeChecker
    display(HTML(html_table.format(content="".join(grid))))

    # add extra dfs to bottom
    [display(list_dfs[i]) for i in range(rows*cols,len(list_dfs))]


if __name__ == "main":
    list_dfs = []
    list_dfs.extend((pd.DataFrame(2*[{"x":"hello"}]),
                     pd.DataFrame(2*[{"x":"world"}]),
                     pd.DataFrame(2*[{"x":"gdbye"}])))

    grid_df_display(3*list_dfs)

In [3]:
from typing import Union

import pandas as pd
from pandas.io.formats.style import Styler


def combination_matrix(dataset: pd.DataFrame, x: str, y: str, z: str,
                       format=None, unique=True) -> Union[pd.DataFrame, Styler]:
    """
    Returns a combination matrix, showing all valid combinations between three DataFrame columns.
    Sort of like a heatmap, but returning lists of (optionally) unique values

    :param dataset: The dataframe to create a combination_matrx from
    :param x: column name to use for the X axis
    :param y: column name to use for the Y axis
    :param z: column name to use for the Z axis (values that appear in the cells)
    :param format: '', ', '-', ', '\n'    = format value lists as "".join() string
                    str, bool, int, float = cast value lists
    :param unique:  whether to return only unique values or not - eg: combination_matrix(unique=False).applymap(sum)
    :return: returns nothing
    """
    unique_y = sorted(dataset[y].unique())
    combinations = pd.DataFrame({
        n: dataset.where(lambda df: df[y] == n)
            .groupby(x)[z]
            .pipe(lambda df: df.unique() if unique else df )
            .apply(list)
            .apply(sorted)
        for n in unique_y
    }).T

    if isinstance(format, str):
        combinations = combinations.applymap(
            lambda cell: f"{format}".join([str(value) for value in list(cell) ])
            if isinstance(cell, list) else cell
        )
    if format == str:   combinations = combinations.applymap(lambda cell: str(cell)      if isinstance(cell, list) and len(cell) > 0 else ''     )
    if format == bool:  combinations = combinations.applymap(lambda cell: True           if isinstance(cell, list) and len(cell) > 0 else False  )
    if format == int:   combinations = combinations.applymap(lambda cell: int(cell[0])   if isinstance(cell, list) and len(cell)     else ''     )
    if format == float: combinations = combinations.applymap(lambda cell: float(cell[0]) if isinstance(cell, list) and len(cell)     else ''     )

    combinations.index.rename(y, inplace=True)
    combinations.fillna('', inplace=True)
    if format == '\n':
        return combinations.style.set_properties(**{'white-space': 'pre-wrap'})  # needed for display
    else:
        return combinations  # Allows for subsequent .applymap()

## Inspect Raw Data

In [4]:
!ls ../input/bengaliai-cv19/

class_map.csv		   test_image_data_3.parquet
sample_submission.csv	   train.csv
test.csv		   train_image_data_0.parquet
test_image_data_0.parquet  train_image_data_1.parquet
test_image_data_1.parquet  train_image_data_2.parquet
test_image_data_2.parquet  train_image_data_3.parquet


In [5]:
dataset = pd.read_csv('../input/bengaliai-cv19/train.csv'); 
# for key in ['grapheme_root','vowel_diacritic','consonant_diacritic','grapheme']:
#     dataset[key] = dataset[key].astype('category')  # ensures groupby().count() shows zeros
dataset.head()

image_id  grapheme_root  vowel_diacritic  consonant_diacritic grapheme
0  Train_0             15                9                    5   ক্ট্রো
1  Train_1            159                0                    0        হ
2  Train_2             22                3                    5     খ্রী
3  Train_3             53                2                    2     র্টি
4  Train_4             71                9                    5     থ্রো

## Question: How many unique graphemes are there?

There are 168 grapheme roots, 11 vowel diacritics, 7 consonant diacritics, and 1295 unique graphemes within the 20k training dataset. 

In [6]:
unique = dataset.apply(lambda col: col.nunique()); unique

image_id               200840
grapheme_root             168
vowel_diacritic            11
consonant_diacritic         7
grapheme                 1295
dtype: int64

## Question: Can all diacritics be used with any grapheme?

- Documentation claims 10,000+ possible graphemes, which is indeed `168 * 11 * 7 = 12936`

- Assuming that the training dataset is representative of common usage, 
  then certian combinations may never (or rarely) be used in practice).

- Unconfirmed Theory: the physics of the human mouth may make such combinations unpronouncable.

- Conclusion: it may be able infer excluded combinations using simple logical rules

### Vowel / Consonant Combinations:
- Vowel #0 and Consonant #0 combine with everything
- Vowels #3, #5, #6, #8 have limited combinations with Consonants 
- Consonant #3 is never combined except with Vowel #0
- Consonant #6 only combineds with Vowels #0 and #1

In [7]:
combination_matrix(dataset, x='consonant_diacritic', y='vowel_diacritic', z='consonant_diacritic', unique=False).applymap(len)

consonant_diacritic    0.0   1.0   2.0  3.0   4.0   5.0   6.0
vowel_diacritic                                              
0                    23960   768  6262  619  5413  4180   306
1                    18799  2843  3838    0  6573  3752  1081
2                    17449   464  3764    0  1255  3035     0
3                    11391     0  2290    0     0  2471     0
4                    11832  1215  1563    0  2206  2032     0
5                     3794     0   297    0   784   422     0
6                     3873     0   463    0     0     0     0
7                    16991  1197  3778    0  4072  2685     0
8                     3210     0     0    0   167   151     0
9                    10727   774  1210    0   800  2521     0
10                    3252   163     0    0     0   148     0

### Grapheme Root Combinations:
- Vowel #0 and Consonant #0 combine with (nearly) everything
- ALL Roots combine with some Consonant #0
- Several Roots do NOT combine with Vowel #0 = [26, 28, 33, 34, 73, 82, 108, 114, 126, 152, 157, 158, 163]
- Several Roots do combine ALL Vowels = [13, 23, 64, 72, 79, 81, 96, 107, 113, 115, 133, 147]}
- Only Root #107 combines with ALL Consonants

In [8]:
root_vowels            = dataset.groupby('grapheme_root')['vowel_diacritic'].unique().apply(sorted).to_frame().T
root_consonants        = dataset.groupby('grapheme_root')['consonant_diacritic'].unique().apply(sorted).to_frame().T
root_vowels_values     = root_vowels.applymap(len).values.flatten()
root_consonants_values = root_consonants.applymap(len).values.flatten()

display(root_vowels)
display({
    "mean":   root_vowels_values.mean(),
    "median": np.median( root_vowels_values ),
    "min":    root_vowels_values.min(),
    "max":    root_vowels_values.max(),
    "unique_vowels":    unique['vowel_diacritic'],
    "root_combine_0":   sum([ 0 in lst for lst in root_vowels.values.flatten() ]),
    "unique_roots":     unique['grapheme_root'],
    "root_combine_not_0": str([ index for index, lst in enumerate(root_vowels.values.flatten()) if 0 not in lst ]),    
    "root_combine_all":       [ index for index, lst in enumerate(root_vowels.values.flatten()) if len(lst) == unique['vowel_diacritic'] ],
})
# print('--------------------')
display(root_consonants)
display({
    "mean":   root_consonants_values.mean(),
    "median": np.median( root_consonants_values ),
    "min":    root_consonants_values.min(),
    "max":    root_consonants_values.max(),
    "unique_consonants":  unique['consonant_diacritic'],
    "root_combine_0": sum([ 0 in lst for lst in root_consonants.values.flatten() ]),
    "unique_roots":   unique['grapheme_root'],
    "root_combine_not_0": str([ index for index, lst in enumerate(root_consonants.values.flatten()) if 0 not in lst ]),        
    "root_combine_all":       [ index for index, lst in enumerate(root_consonants.values.flatten()) if len(lst) == unique['consonant_diacritic'] ],
})

grapheme_root    0    1       2    3    4    5    6    7    8       9    10   \
vowel_diacritic  [0]  [0]  [0, 1]  [0]  [0]  [0]  [0]  [0]  [0]  [0, 1]  [0]   

grapheme_root    11   12                                  13   \
vowel_diacritic  [0]  [0]  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]   

grapheme_root                14               15                  16   \
vowel_diacritic  [0, 1, 2, 7, 9]  [0, 1, 2, 7, 9]  [0, 1, 2, 6, 7, 9]   

grapheme_root                17                         18       19      20   \
vowel_diacritic  [0, 1, 2, 7, 9]  [0, 1, 2, 3, 4, 7, 9, 10]  [0, 10]  [0, 3]   

grapheme_root                21                           22   \
vowel_diacritic  [0, 1, 2, 3, 7]  [0, 1, 2, 3, 4, 6, 7, 8, 9]   

grapheme_root                                   23      24   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  [0, 1]   

grapheme_root                   25   26      27            28   \
vowel_diacritic  [0, 1, 2, 3, 4, 7]  [7]  [0, 2]  [1, 2, 4, 9]   

grapheme_root                         29         30               31   \
vowel_diacritic  [0, 1, 2, 4, 5, 6, 7, 9]  [0, 2, 7]  [0, 1, 2, 4, 7]   

grapheme_root                         32   33      34         35   \
vowel_diacritic  [0, 1, 2, 3, 4, 6, 7, 9]  [2]  [1, 2]  [0, 1, 2]   

grapheme_root                      36      37   \
vowel_diacritic  [0, 1, 2, 3, 4, 7, 9]  [0, 2]   

grapheme_root                                38            39   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 7, 8, 9, 10]  [0, 1, 2, 4]   

grapheme_root                40      41                  42   \
vowel_diacritic  [0, 1, 2, 4, 7]  [0, 1]  [0, 1, 2, 4, 7, 9]   

grapheme_root                         43                     44   45   \
vowel_diacritic  [0, 1, 2, 3, 4, 7, 8, 9]  [0, 1, 2, 3, 4, 7, 9]  [0]   

grapheme_root             46         47                  48      49   \
vowel_diacritic  [0, 1, 3, 7]  [0, 1, 2]  [0, 1, 2, 4, 7, 9]  [0, 1]   

grapheme_root                50      51                  52   \
vowel_diacritic  [0, 1, 2, 4, 7]  [0, 2]  [0, 1, 2, 3, 4, 7]   

grapheme_root                         53            54   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 7, 9]  [0, 1, 2, 9]   

grapheme_root                      55                     56            57   \
vowel_diacritic  [0, 1, 2, 3, 4, 7, 9]  [0, 1, 2, 3, 4, 7, 9]  [0, 1, 2, 3]   

grapheme_root                       58                     59         60   \
vowel_diacritic  [0, 1, 2, 4, 7, 9, 10]  [0, 1, 2, 3, 4, 7, 9]  [0, 1, 4]   

grapheme_root             61                  62   63   \
vowel_diacritic  [0, 1, 2, 7]  [0, 1, 2, 3, 4, 7]  [0]   

grapheme_root                                   64                  65   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  [0, 1, 2, 3, 7, 9]   

grapheme_root                66         67            68               69   \
vowel_diacritic  [0, 1, 2, 3, 7]  [0, 1, 7]  [0, 2, 3, 7]  [0, 1, 2, 3, 7]   

grapheme_root                70                        71   \
vowel_diacritic  [0, 1, 2, 3, 7]  [0, 1, 2, 3, 4, 7, 8, 9]   

grapheme_root                                   72   73   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  [1]   

grapheme_root                          74               75   \
vowel_diacritic  [0, 1, 2, 3, 4, 7, 9, 10]  [0, 1, 2, 6, 7]   

grapheme_root                            76               77      78   \
vowel_diacritic  [0, 1, 2, 3, 4, 6, 7, 8, 9]  [0, 1, 2, 4, 5]  [0, 1]   

grapheme_root                                   79      80   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  [0, 7]   

grapheme_root                                   81      82   \
vowel_diacritic  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]  [3, 4]   

grapheme_root                   83         84                     85   \
vowel_diacritic  [0, 1, 2, 4, 7, 9]  [0, 2, 7]  [0, 1, 2, 3, 4, 7, 9]   

grapheme_root                      86   87               88   \
vowel_diacritic  [0, 1, 2, 3, 4, 7, 9]  [0]  [0, 1, 2, 3, 7]   

grapheme_root                  

{'mean': 4.869047619047619,
 'median': 5.0,
 'min': 1,
 'max': 11,
 'unique_vowels': 11,
 'root_combine_0': 155,
 'unique_roots': 168,
 'root_combine_not_0': '[26, 28, 33, 34, 73, 82, 108, 114, 126, 152, 157, 158, 163]',
 'root_combine_all': [13, 23, 64, 72, 79, 81, 96, 107, 113, 115, 133, 147]}

grapheme_root        0    1       2       3       4    5       6    7    8    \
consonant_diacritic  [0]  [0]  [0, 4]  [0, 1]  [0, 1]  [0]  [0, 1]  [0]  [0]   

grapheme_root              9    10   11   12                  13   14   \
consonant_diacritic  [0, 1, 4]  [0]  [0]  [0]  [0, 1, 2, 4, 5, 6]  [0]   

grapheme_root           15   16   17      18   19   20      21   \
consonant_diacritic  [0, 5]  [0]  [0]  [0, 4]  [0]  [0]  [0, 2]   

grapheme_root                    22                  23   24      25   26   \
consonant_diacritic  [0, 1, 2, 4, 5]  [0, 1, 2, 4, 5, 6]  [0]  [0, 4]  [0]   

grapheme_root        27      28                  29   30   31   32   33   34   \
consonant_diacritic  [0]  [0, 4]  [0, 1, 2, 3, 4, 5]  [0]  [0]  [0]  [0]  [0]   

grapheme_root        35   36   37            38   39   40   41            42   \
consonant_diacritic  [0]  [0]  [0]  [0, 1, 2, 4]  [0]  [0]  [0]  [0, 1, 2, 4]   

grapheme_root                    43   44   45   46   47         48   49   50   \
consonant_diacritic  [0, 2, 3, 4, 5]  [0]  [0]  [0]  [0]  [0, 1, 2]  [0]  [0]   

grapheme_root        51   52               53   54         55            56   \
consonant_diacritic  [0]  [0]  [0, 2, 4, 5, 6]  [0]  [0, 1, 4]  [0, 2, 4, 5]   

grapheme_root        57      58         59   60   61   62   63   \
consonant_diacritic  [0]  [0, 4]  [0, 2, 4]  [0]  [0]  [0]  [0]   

grapheme_root                       64         65   66   67   68   69   \
consonant_diacritic  [0, 1, 2, 4, 5, 6]  [0, 2, 4]  [0]  [0]  [0]  [0]   

grapheme_root           70               71                  72   73      74   \
consonant_diacritic  [0, 4]  [0, 2, 3, 4, 5]  [0, 1, 2, 4, 5, 6]  [0]  [0, 2]   

grapheme_root           75         76      77   78               79      80   \
consonant_diacritic  [0, 2]  [0, 2, 4]  [0, 5]  [0]  [0, 1, 2, 4, 5]  [0, 2]   

grapheme_root              81   82      83   84      85            86   87   \
consonant_diacritic  [0, 2, 4]  [0]  [0, 5]  [0]  [0, 5]  [0, 2, 4, 5]  [0]   

grapheme_root        88         89   90         91      92   93   94   95   \
consonant_diacritic  [0]  [0, 4, 5]  [0]  [0, 4, 5]  [0, 4]  [0]  [0]  [0]   

grapheme_root                       96   97   98   99   100     101  102  \
consonant_diacritic  [0, 1, 2, 4, 5, 6]  [0]  [0]  [0]  [0]  [0, 4]  [0]   

grapheme_root                       103  104  105     106  \
consonant_diacritic  [0, 1, 2, 4, 5, 6]  [0]  [0]  [0, 4]   

grapheme_root                          107  108  109  110     111     112  \
consonant_diacritic  [0, 1, 2, 3, 4, 5, 6]  [0]  [0]  [0]  [0, 2]  [0, 4]   

grapheme_root                    113  114           115  116     117  118  \
consonant_diacritic  [0, 1, 2, 4, 5]  [0]  [0, 2, 4, 5]  [0]  [0, 5]  [0]   

grapheme_root           119     120  121        122        123        124  \
consonant_diacritic  [0, 5]  [0, 2]  [0]  [0, 2, 4]  [0, 1, 4]  [0, 2, 4]   

grapheme_root        125  126  127     128  129  130  131  132           133  \
consonant_diacritic  [0]  [0]  [0]  [0, 2]  [0]  [0]  [0]  [0]  [0, 2, 4, 5]   

grapheme_root        134  135     136  137  138        139     140        141  \
consonant_diacritic  [0]  [0]  [0, 2]  [0]  [0]  [0, 2, 4]  [0, 5]  [0, 4, 5]   

grapheme_root           142  143     144  145  146              147  \
consonant_diacritic  [0, 4]  [0]  [0, 5]  [0]  [0]  [0, 1, 2, 4, 5]   

grapheme_root              148              149     150        151  152  \
consonant_diacritic  [0, 4, 5]  [0, 2, 4, 5, 6]  [0, 5]  [0, 2, 4]  [0]   

grapheme_root              153  154  155  156  157  158              159  160  \
consonant_diacritic  [0, 4, 5]  [0]  [0]  [0]  [0]  [0]  [0, 1, 2, 4, 5]  [0]   

grapheme_root        161  162  163  164  165  166     167  
consonant_diacritic  [0]  [0]  [0]  [0]  [0]  [0]  [0, 4]

{'mean': 1.9583333333333333,
 'median': 1.0,
 'min': 1,
 'max': 7,
 'unique_consonants': 7,
 'root_combine_0': 168,
 'unique_roots': 168,
 'root_combine_not_0': '[]',
 'root_combine_all': [107]}

### Combination Matrices

This is the full list of which Grapheme Roots combine with which Vowels and Consonant Diacritics

In [9]:
combination_matrix(dataset, x='consonant_diacritic', y='vowel_diacritic', z='grapheme_root', format=', ')

consonant_diacritic                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  0.0  \
vowel_diacritic                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
0                    0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 27.0, 29.0, 30.0, 31.0, 32.0, 35.0, 36.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 45.0, 46.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 67.0, 68.0, 69.0, 70.0, 71.0, 72.0, 74.0, 75.0, 76.0, 77.0, 78.0, 79.0, 80.0, 81.0, 83.0, 84.0, 85.0, 86.0, 87.0, 88.0, 89.0, 90.0, 91.0, 92.0, 93.0, 94.0, 95.0, 96.0, 97.0, 98.0, 99.0, 100.0, 101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 109.0, 110.0, 111.0, 112.0, 113.0, 115.0, 116.0, 117.0, 118.0, 119.0, 120.0, 121.0, 122.0, 123.0, 124.0, 125.0, 127.0, 128.0, 129.0, 130.0, 131.0, 132.0, 133.0, 134.0, 135.0, 136.0, 137.0, 138.0, 139.0, 140.0, 141.0, 142.0, 143.0, 144.0, 145.0, 146.0, 147.0, 148.0, 149.0, 150.0, 151.0, 153.0, 154.0, 155.0, 156.0, 159.0, 160.0, 161.0, 162.0, 164.0, 165.0, 166.0, 167.0   
1                                                                                                                                                                                                                                13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 21.0, 22.0, 23.0, 24.0, 25.0, 28.0, 29.0, 31.0, 32.0, 34.0, 35.0, 36.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 46.0, 47.0, 48.0, 49.0, 50.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 58.0, 59.0, 60.0, 61.0, 62.0, 64.0, 65.0, 66.0, 67.0, 69.0, 70.0, 71.0, 72.0, 73.0, 74.0, 75.0, 76.0, 77.0, 78.0, 79.0, 81.0, 83.0, 85.0, 86.0, 88.0, 89.0, 91.0, 92.0, 94.0, 95.0, 96.0, 97.0, 98.0, 99.0, 100.0, 101.0, 103.0, 106.0, 107.0, 109.0, 111.0, 112.0, 113.0, 114.0, 115.0, 116.0, 117.0, 118.0, 119.0, 120.0, 121.0, 122.0, 123.0, 124.0, 125.0, 127.0, 128.0, 129.0, 132.0, 133.0, 134.0, 135.0, 136.0, 139.0, 140.0, 141.0, 142.0, 143.0, 144.0, 147.0, 148.0, 149.0, 150.0, 151.0, 152.0, 153.0, 155.0, 156.0, 157.0, 159.0, 161.0, 162.0, 163.0, 

### Visualizing Bengali

In [10]:
combination_matrix(dataset, x='consonant_diacritic', y='vowel_diacritic', z='grapheme', format=' ')

consonant_diacritic                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             0.0  \
vowel_diacritic                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
0                           ং ঃ অ আ ই ঈ উ ঊ ঋ এ ঐ ও ঔ ক ক্ক ক্ট ক্ত ক্ল ক্ষ ক্ষ্ণ ক্ষ্ম ক্স খ গ গ্ধ গ্ন গ্ম ঘ ঘ্ন ঙ ঙ্ক ঙ্খ ঙ্গ ঙ্ঘ চ চ্চ চ্ছ চ্ছ্ব ছ জ জ্জ জ্জ্ব জ্ঞ জ্ব ঝ ঞ ঞ্চ ঞ্ছ ঞ্জ ট ট্ট ঠ ড ড্ড ঢ ণ ণ্ট ণ্ঠ ণ্ড ণ্ণ ত ত্ত ত্ত্ব ত্থ ত্ন ত্ব ত্ম থ দ দ্দ দ্ধ দ্ব দ্ভ দ্ম ধ ধ্ব ন ন্ট ন্ঠ ন্ড ন্ত ন্ত্ব ন্থ ন্দ ন্দ্ব ন্ধ ন্ন ন্ব ন্ম ন্স প প্ট প্ত প্ন প্প প্ল প্স ফ ফ্ট ফ্ফ ফ্ল ব ব্দ ব্ধ ব্ব ব্ল ভ ম ম্ন ম্প ম্ব ম্ভ ম্ম ম্ল য র ল ল্ক ল্ট ল্ড ল্প ল্ব ল্ম ল্ল শ শ্চ শ্ন শ্ব শ্ম শ্ল ষ ষ্ক ষ্ট ষ্ঠ ষ্ণ ষ্প ষ্ফ ষ্ম স স্ক স্ট স্ত স্থ স্প স্ফ স্ব স্ম হ হ্ন হ্ব হ্ম ৎ ড় ঢ় য়   
1                    কা ক্কা ক্টা ক্তা ক্লা ক্ষা ক্সা খা গা গ্ধা গ্না গ্লা ঘা ঙা ঙ্কা ঙ্ক্ষা ঙ্খা ঙ্গা চা চ্চা চ্ছা চ্ছ্বা ছা জা জ্জা জ্ঞা জ্বা ঝা ঞা ঞ্চা ঞ্জা টা ট্টা ঠা ডা ড্ডা ঢা ণা ণ্টা ণ্ঠা ণ্ডা তা ত্তা ত্ত্বা ত্থা ত্বা ত্মা থা দা দ্ঘা দ্দা দ্ধা দ্বা দ্ভা দ্মা ধা না ন্টা ন্ডা ন্তা ন্থা ন্দা ন্ধা ন্না ন্মা ন্সা পা প্টা প্তা প্না প্পা প্লা ফা ফ্লা বা ব্দা ব্বা ব্লা ভা ভ্লা মা ম্না ম্পা ম্বা ম্ভা ম্মা ম্লা যা রা লা ল্কা ল্টা ল্ডা ল্পা ল্লা শা শ্চা শ্না শ্বা ষা ষ্কা ষ্টা ষ্ঠা ষ্ণা ষ্পা সা স্কা স্টা স্তা স্থা স্না স্পা স্বা স্মা স্লা হা হ্বা হ্মা হ্লা ড়া য়া   
2                                                        কি ক্কি ক্টি ক্তি ক্লি ক্ষি ক্সি খি গি গ্নি গ্মি গ্লি ঘি ঘ্নি ঙি ঙ্কি ঙ্ক্তি ঙ্ক্ষি ঙ্খি ঙ্গি ঙ্ঘি চি চ্চি চ্ছি ছি জি জ্জি জ্বি ঝি ঞ্চি ঞ্ছি ঞ্জি টি ট্টি ঠি ডি ড্ডি ঢি ণি ণ্ঠি ণ্ডি তি ত্তি ত্ত্বি ত্নি ত্বি ত্মি থি দি দ্দি দ্ধি দ্বি দ্ভি ধি নি ন্টি ন্ঠি ন্ডি ন্তি ন্থি ন্দি ন্দ্বি ন্ধি ন্নি ন্বি ন্মি ন্সি পি প্তি প্লি ফি বি ব্জি ব্দি ব্ধি ব্বি ব্লি ভি মি ম্পি ম্বি ম্ভি ম্মি যি রি লি ল্কি ল্টি ল্ডি ল্পি ল্মি ল্লি শি শ্চি শ্বি শ্মি শ্লি ষি ষ্টি ষ্ঠি সি স্কি স্টি স্তি স্থি স্নি স্পি স্বি স্মি হি হ্নি ড়ি য়ি   
3                                                                                                                                                                                                                                              কী ক্ষী ক্ষ্মী খী গী গ্নী ঙ্কী ঙ্গী চী জী জ্জী জ্ঞী ঞ্জী টী ঠী ডী ড্ডী ণী ণ্ডী তী ত্তী ত্ত্বী ত্নী ত্বী ত্মী থী দী দ্দী দ্বী ধী নী ন্জী ন্ডী ন্তী ন্থী ন্দী ন্দ্বী ন্ধী ন্নী ন্বী ন্সী পী প্পী ফী বী ব্দী ভী মী ম্নী ম্বী ম্ভী ম্মী যী রী লী ল্কী ল্পী ল্লী শী শ্মী শ্লী ষী ষ্টী ষ্ঠী সী স্টী স্তী স্পী স্ফী স্বী স্মী হী হ্মী ড়ী য়ী   
4                                                                                                                                                                                                                                                         কু ক্ষু খু গু গ্লু ঘু ঙু ঙ্কু ঙ্গু চু চ্চু চ্ছু ছু জু জ্জু ঝু ঞ্চু ঞ্জু টু ঠু ডু ঢু ণু ণ্টু ণ্ডু তু থু দু দ্দু দ্বু দ্ভু ধু নু ন্জু ন্টু ন্ডু ন্তু ন্দু ন্ধু ন্নু ন্মু পু প্লু ফু বু ব্দু ব্বু ব্লু ভু মু ম্পু ম্বু ম্ভু ম্মু যু রু লু ল্গু ল্টু ল্লু শু শ্চু শ্লু ষু ষ্টু ষ্ঠু ষ্ণু সু স্কু স্টু স্তু স্ফু স্সু হু ড়ু য়ু   
5                              

In [11]:
combination_matrix(dataset, x='grapheme_root', y='vowel_diacritic', z='grapheme', format=' ')

grapheme_root   0.0   1.0   2.0   3.0   4.0   5.0   6.0   7.0   8.0   9.0    \
vowel_diacritic                                                               
0                   ং     ঃ     অ  আ আঁ  ই ইঁ     ঈ  উ উঁ     ঊ     ঋ  এ এঁ   
1                            অ্যা                                      এ্যা   
2                                                                             
3                                                                             
4                                                                             
5                                                                             
6                                                                             
7                                                                             
8                                                                             
9                                                                             
10                                                                            

grapheme_root   10.0  11.0  12.0                          13.0  14.0   \
vowel_diacritic                                                         
0                   ঐ     ও     ঔ                 ক ক্য ক্র র্ক   ক্ক   
1                                  কা কাঁ ক্যা ক্রা ক্র্যা র্কা  ক্কা   
2                                                  কি ক্রি র্কি  ক্কি   
3                                                  কী ক্রী র্কী         
4                                         কু কুঁ ক্যু ক্রু র্কু         
5                                                       কূ ক্রূ         
6                                                            কৃ         
7                                         কে কেঁ ক্যে ক্রে র্কে  ক্কে   
8                                                            কৈ         
9                                                  কো ক্রো র্কো  ক্কো   
10                                                           কৌ         

grapheme_root          15.0  16.0  17.0         18.0    19.0    20.0   \
vowel_diacritic                                                         
0                  ক্ট ক্ট্র   ক্ত   ক্ল    ক্ষ ক্ষ্য   ক্ষ্ণ   ক্ষ্ম   
1                       ক্টা  ক্তা  ক্লা  ক্ষা ক্ষ্যা                   
2                       ক্টি  ক্তি  ক্লি         ক্ষি                   
3                                                ক্ষী          ক্ষ্মী   
4                                                ক্ষু                   
5                                                                       
6                             ক্তৃ                                      
7                       ক্টে  ক্তে  ক্লে  ক্ষে ক্ষ্যে                   
8                                                                       
9                ক্টো ক্ট্রো  ক্তো  ক্লো         ক্ষো                   
10                                               ক্ষৌ  ক্ষ্ণৌ           

grapheme_root        21.0              22.0                          23.0   \
vowel_diacritic                                                              
0                ক্স র্ক্স         খ খ্য র্খ                 গ গ্য গ্র র্গ   
1                     ক্সা  খা খাঁ খ্যা র্খা  গা গাঁ গ্যা গ্রা গ্র্যা র্গা   
2                     ক্সি           খি খ্রি             গি গ্যি গ্রি র্গি   
3                   র্ক্সী           খী খ্রী                  গী গ্রী র্গী   
4                                     খু খুঁ              গু গুঁ গ্রু র্গু   
5                                                                       গূ   
6                                         খৃ                            গৃ   
7                     ক্সে           খে র্খে         গে গেঁ গ্যে গ্রে র্গে   
8                                         খৈ                            গৈ   
9                                     খো খোঁ              গো গোঁ গ্রো র্গো   
10                                                                      গৌ   

grapheme_root   24.0         25.0  26.0  27.0         28.

In [12]:
combination_matrix(dataset, x='grapheme_root', y='consonant_diacritic', z='grapheme', format=' ')

grapheme_root       0.0   1.0   2.0   3.0   4.0   5.0   6.0   7.0   8.0    \
consonant_diacritic                                                         
0                       ং     ঃ     অ     আ     ই     ঈ     উ     ঊ     ঋ   
1                                        আঁ    ইঁ          উঁ               
2                                                                           
3                                                                           
4                                অ্যা                                       
5                                                                           
6                                                                           

grapheme_root       9.0   10.0  11.0  12.0   \
consonant_diacritic                           
0                       এ     ঐ     ও     ঔ   
1                      এঁ                     
2                                             
3                                             
4                    এ্যা                     
5                                             
6                                             

grapheme_root                                         13.0   \
consonant_diacritic                                           
0                           ক কা কি কী কু কূ কৃ কে কৈ কো কৌ   
1                                               কাঁ কুঁ কেঁ   
2                         র্ক র্কা র্কি র্কী র্কু র্কে র্কো   
3                                                             
4                                        ক্য ক্যা ক্যু ক্যে   
5                    ক্র ক্রা ক্রি ক্রী ক্রু ক্রূ ক্রে ক্রো   
6                                                    ক্র্যা   

grapheme_root                          14.0                     15.0   \
consonant_diacritic                                                     
0                    ক্ক ক্কা ক্কি ক্কে ক্কো  ক্ট ক্টা ক্টি ক্টে ক্টো   
1                                                                       
2                                                                       
3                                                                       
4                                                                       
5                                                        ক্ট্র ক্ট্রো   
6                                                                       

grapheme_root                               16.0                     17.0   \
consonant_diacritic                                                          
0                    ক্ত ক্তা ক্তি ক্তৃ ক্তে ক্তো  ক্ল ক্লা ক্লি ক্লে ক্লো   
1                                                                            
2                                                                            
3                                                                            
4                                                                            
5                                                                            
6                                                                            

grapheme_root                                         18.0          19.0   \
consonant_diacritic                                                         
0                    ক্ষ ক্ষা ক্ষি ক্ষী ক্ষু ক্ষে ক্ষো ক্ষৌ  ক্ষ্ণ ক্ষ্ণৌ   
1                                                                           
2                                                                           
3                                                                           
4                                       ক্ষ্য ক্ষ্যা ক্ষ্যে                 
5                                                                           
6                                                                           

grapheme_root               20.0                21.0   \
consonant_diacritic                                     
0                    ক্ষ্ম ক্ষ্মী  ক্স ক্সা ক্সি ক্সে   
1                                                       
2                              

## Sanity Checking = Found Dataset BUG!

This combination_matrix lists 1292 unique grapheme combinations, which is 3 less than the 1295 unique graphemes listed in the training dataset. Something is WRONG!

Found a discrepency BUG is in the dataset. The following root/vowel/consonant keys have multiple unicode graphemes renderings! 

{'64-3-2': ['র্তী', 'র্ত্রী'],
 '64-7-2': ['র্তে', 'র্ত্রে'],
 '72-0-2': ['র্দ্র', 'র্দ']}

In [13]:
from itertools import chain
{
    "combinations": len(list(chain( 
        *combination_matrix(dataset, x='consonant_diacritic', y='vowel_diacritic', z='grapheme_root')
        .values.flatten() 
    ))),
    "unique_graphemes": unique['grapheme']
}

{'combinations': 1292, 'unique_graphemes': 1295}

- Confirm that there are no null or NaN values in the dataset

In [14]:
dataset.apply(lambda row: row.isnull()).sum()

image_id               0
grapheme_root          0
vowel_diacritic        0
consonant_diacritic    0
grapheme               0
dtype: int64

- Found the BUG! It is in the dataset!
- There is THREE sets of unique root/vowel/consonant keys that have multiple unicode renderings 

In [15]:
( 
    dataset
    .groupby(['grapheme_root', 'vowel_diacritic', 'consonant_diacritic'])
    .nunique(dropna=False) > 1 
).sum()

image_id               1292
grapheme_root             0
vowel_diacritic           0
consonant_diacritic       0
grapheme                  3
dtype: int64

In [16]:
( 
    dataset
    .groupby(['grapheme_root', 'vowel_diacritic', 'consonant_diacritic'])
    .nunique(dropna=False) > 1
).query("grapheme != False")

image_id  grapheme_root  \
grapheme_root vowel_diacritic consonant_diacritic                            
64            3               2                        True          False   
              7               2                        True          False   
72            0               2                        True          False   

                                                   vowel_diacritic  \
grapheme_root vowel_diacritic consonant_diacritic                    
64            3               2                              False   
              7               2                              False   
72            0               2                              False   

                                                   consonant_diacritic  \
grapheme_root vowel_diacritic consonant_diacritic                        
64            3               2                                  False   
              7               2                                  False   
72            0               2                                  False   

                                                   grapheme  
grapheme_root vowel_diacritic consonant_diacritic            
64            3               2                        True  
              7               2                        True  
72            0               2                        True

In [17]:
multilabled_graphemes = {
    "64-3-2": dataset.query("grapheme_root == 64 & vowel_diacritic == 3 & consonant_diacritic == 2")['grapheme'].unique().tolist(),
    "64-7-2": dataset.query("grapheme_root == 64 & vowel_diacritic == 7 & consonant_diacritic == 2")['grapheme'].unique().tolist(),
    "72-0-2": dataset.query("grapheme_root == 72 & vowel_diacritic == 0 & consonant_diacritic == 2")['grapheme'].unique().tolist(),
}
multilabled_graphemes

{'64-3-2': ['র্তী', 'র্ত্রী'],
 '64-7-2': ['র্তে', 'র্ত্রে'],
 '72-0-2': ['র্দ্র', 'র্দ']}

As discovered by @ren4yu[[1](https://www.kaggle.com/c/bengaliai-cv19/discussion/133735#763562)] the unicode itself is encoded using the root/vowel/consonant diacritics as a multibyte string, and the unicode consortium have implemented multiple renderings of the same grapheme by allowing duplicate diacritics within the unicode. 

This potentually opens up another datasource for investigation, which is to explore the full range of diacritic combinations within the unicode specification.

The paper [Fonts-2-Handwriting: A Seed-Augment-Train framework for universal digit classification](https://arxiv.org/pdf/1905.08633.pdf) also makes the suggestion that it may be possible to generate synethetic data for handwriting recognition by rendering each of the unicode graphemes using various Bengali fonts

This simply counts how many times each of these multi-keyed unicode graphemes is listed in the database

In [18]:
dataset[ dataset['grapheme'].isin(multilabled_grapheme_list) ].groupby(['grapheme']).count()['image_id'].to_dict()


NameError: name 'multilabled_grapheme_list' is not defined